zebra-chain/src/block/merkle.rs

```diff
@@ -99,18 +121,59 @@ impl std::iter::FromIterator<transaction::Hash> for Root {
         I: IntoIterator<Item = transaction::Hash>,
     {
         let mut hashes = hashes.into_iter().map(|hash| hash.0).collect::<Vec<_>>();
+        root(&mut hashes);
+        Self(hashes[0])
+    }
+}
 
-        while hashes.len() > 1 {
-            hashes = hashes
-                .chunks(2)
-                .map(|chunk| match chunk {
-                    [h1, h2] => hash(h1, h2),
-                    [h1] => hash(h1, h1),
-                    _ => unreachable!("chunks(2)"),
-                })
-                .collect();
-        }
+/// The root of the authorizing data Merkle tree, binding the
+/// block header to the authorizing data of the block (signatures, proofs)
+/// as defined in [ZIP-244].
+///
+/// See [`Root`] for an important disclaimer.
+///
+/// [ZIP-244]: https://zips.z.cash/zip-0244
+#[derive(Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub struct AuthDataRoot(pub [u8; 32]);
+
+impl fmt::Debug for AuthDataRoot {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("AuthRoot")
+            .field(&hex::encode(&self.0))
+            .finish()
+    }
+}
+
+impl<T> std::iter::FromIterator<T> for AuthDataRoot
+where
+    T: std::convert::AsRef<Transaction>,
+{
+    fn from_iter<I>(transactions: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+    {
+        // > For transaction versions before v5, a placeholder value consisting
+        // > of 32 bytes of 0xFF is used in place of the authorizing data commitment.
+        // > This is only used in the tree committed to by hashAuthDataRoot.
+        transactions
+            .into_iter()
+            .map(|tx| {
+                tx.as_ref()
+                    .auth_digest()
+                    .unwrap_or_else(|| transaction::AuthDigest([0xFF; 32]))
+            })
+            .collect()
+    }
+}
 
+impl std::iter::FromIterator<transaction::AuthDigest> for AuthDataRoot {
+    fn from_iter<I>(hashes: I) -> Self
+    where
+        I: IntoIterator<Item = transaction::AuthDigest>,
+    {
+        let mut hashes = hashes.into_iter().map(|hash| hash.0).collect::<Vec<_>>();
+        root(&mut hashes);
```

Comment:
Same here, and what @teor2345 said about the different paddings.

I would be fine with having two `root()` functions, something like `hash_merkle_root()` and `hash_auth_data_root()`, since they are themselves small. Maybe a general one that takes the padding function or param, and then each of those are just calls to the underlying function.